packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -232,10 +255,29 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
       if (allowsCustomValue && inputValue !== itemText) {
         commitCustomValue();
       } else {
-        resetInputValue();
-        // Close menu if blurring away from the combobox
-        // Specifically handles case where user clicks away from the field
-        triggerState.close();
+        // If multiple things are controlled, call onSelectionChange
+        // Note that users who control open state and input value MUST also have a onSelectionChange handler to properly close the menu and reset input value
+        if (
+          (props.selectedKey !== undefined && props.inputValue !== undefined) ||
+          (props.inputValue !== undefined && props.isOpen !== undefined) ||
+          (props.selectedKey !== undefined && props.isOpen !== undefined)
+        ) {
+          props.onSelectionChange(selectedKey);
+
+          // If multiple things are controlled but inputValue isn't, reset the input value for the user
+          if (props.inputValue === undefined) {
+            resetInputValue();
+          }
+
+          // If multiple things are controlled but isOpen isn't, close the menu for the user
+          if (props.isOpen === undefined) {
+            triggerState.close();
+          }
+        } else {
+          // If only a single aspect of combobox is controlled, reset input value and close menu for the user
+          resetInputValue();
+          triggerState.close();
+        }
```

Comment:
The above is very similar to this. Do you think we should combine them into a function that we can reuse?